From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: OverwriteMC <artyom030803@gmail.com>
Date: Sat, 28 Feb 2026 14:17:02 +0900
Subject: [PATCH] Optimize pushable selector

Avoid duplicated Bukkit#isPushable check, already checked inside the Bukkit#canCollideWithBukkit

diff --git a/net/minecraft/world/entity/EntitySelector.java b/net/minecraft/world/entity/EntitySelector.java
index 326da0817707fb2cd145c2c387d88b5e6ad780d4..409ffce06e0e38d7e98302a0171b6af2c9603941 100644
--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -46,10 +46,9 @@ public final class EntitySelector {
         Team team = entity.getTeam();
         Team.CollisionRule collisionRule = team == null ? Team.CollisionRule.ALWAYS : team.getCollisionRule();
         return (Predicate<Entity>)(collisionRule == Team.CollisionRule.NEVER
-            ? Predicates.alwaysFalse()
-            : NO_SPECTATORS.and(
-                entity1 -> { final Entity pushedEntity = entity1; // Paper - OBFHELPER
-                    if (!pushedEntity.isPushable() || !pushedEntity.canCollideWithBukkit(entity) || !entity.canCollideWithBukkit(pushedEntity)) { // CraftBukkit - collidable API // Paper - Climbing should not bypass cramming gamerule
+            ? Predicates.<Entity>alwaysFalse() // Leaf - optimize pushable selector
+            : (Predicate<Entity>) (entity1) -> { final Entity pushedEntity = entity1; // Paper - OBFHELPER // Leaf - optimize pushable selector
+                    if (entity1.isSpectator() /*|| !pushedEntity.isPushable()*/ || !pushedEntity.canCollideWithBukkit(entity) || !entity.canCollideWithBukkit(pushedEntity)) { // CraftBukkit - collidable API // Paper - Climbing should not bypass cramming gamerule // Leaf - optimize pushable selector
                         return false;
                     } else if (!entity.level().isClientSide() || entity1 instanceof Player player && player.isLocalPlayer()) {
                         Team team1 = entity1.getTeam();
@@ -65,7 +64,7 @@ public final class EntitySelector {
                         return false;
                     }
                 }
-            ));
+            ); // Leaf - optimize pushable selector
     }
 
     public static Predicate<Entity> notRiding(Entity entity) {
